graphql-kotlin-spring-server/src/main/kotlin/com/expediagroup/graphql/spring/execution/ApolloSubscriptionSessionState.kt

```diff
@@ -53,16 +55,19 @@ internal class ApolloSubscriptionSessionState {
     /**
      * Stop the subscription sending data. Does NOT terminate the session.
      */
-    fun stopOperation(session: WebSocketSession, operationMessage: SubscriptionOperationMessage) {
+    fun stopOperation(session: WebSocketSession, operationMessage: SubscriptionOperationMessage): Flux<SubscriptionOperationMessage> {
         if (operationMessage.id != null) {
             val operationsForSession = activeOperations[session.id]
-            operationsForSession?.get(operationMessage.id)?.cancel()
-            operationsForSession?.remove(operationMessage.id)
-
-            if (operationsForSession?.isEmpty() == true) {
-                activeOperations.remove(session.id)
+            operationsForSession?.get(operationMessage.id)?.let {
+                it.cancel()
+                operationsForSession.remove(operationMessage.id)
+                if (operationsForSession.isEmpty()) {
+                    activeOperations.remove(session.id)
+                }
+                return Flux.just(SubscriptionOperationMessage(type = GQL_COMPLETE.type, id = operationMessage.id))
```

Comment:
This behaviour is not correct. We need to send the complete event back to clients when the subscription is complete, not when the operation has an error or the client manually stops it

https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md#gql_complete

Reply:
For queries and mutation yes, for subscriptions this is different:
`For subscriptions:
App triggers PubSub's publication method, and the server publishes the event, passing it through the subscribe executor to create GraphQL execution result
Client receives GQL_DATA with the data, and handles it.
When client unsubscribe, the server triggers onOperationDone and sends GQL_COMPLETE message to the client.`
Is it the intent to also handle queries and mutation over websocket? Cause for graphql-java-kickstart this is not in scope.

Reply:
NVM from the spec as of December 2019

> Server sends GQL_COMPLETE if the operation is a query or mutation (for subscriptions, this sent when unsubscribing)